clippy_lints/src/identity_op.rs

```diff
@@ -31,33 +32,99 @@ declare_clippy_lint! {
 declare_lint_pass!(IdentityOp => [IDENTITY_OP]);
 
 impl<'tcx> LateLintPass<'tcx> for IdentityOp {
-    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) {
-        if e.span.from_expansion() {
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &'tcx Body<'_>) {
+        if body.value.span.from_expansion() {
             return;
         }
-        if let ExprKind::Binary(cmp, left, right) = e.kind {
-            if is_allowed(cx, cmp, left, right) {
-                return;
+        let mut visitor = ExprVisitor {
+            cx,
+            leftmost: true,
+            top_level_binop: true,
+        };
+        visitor.visit_expr(&body.value);
+    }
+}
+
+struct ExprVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    leftmost: bool,
+    top_level_binop: bool,
+}
```

Comment:
I think a visitor is too confusing here, e.g. `leftmost` is always true by the time it reaches the `self.reducible_to_right` calls

A simpler alternative would be to go back to using `check_expr` in the lint pass and checking if `e`'s [parent expr](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.get_parent_expr.html) is an `ExprKind::Binary` to determine if it's top level or not

Reply:
Thanks for your review. I fixed it.

Reply:
I think the visitor can be removed entirely, at least for all the test cases `leftmost` is still always true in `self.reducible_to_right` because the order statements are visited in are

```rust
     a + b + c + d;
// 1 LLLLLLLLL   R
// 2 LLLLL   R
// 3 L   R
```

I believe just checking the `ExprKind` of right as you do in `reducible_to_right` + `is_toplevel_binary` is enough

Reply:
Ohh I misunderstood what you meant. And it seems to work for test cases.

Reply:
Fixed it.